lib/Sema/CSSimplify.cpp

```diff
@@ -7985,6 +7995,10 @@ ConstraintSystem::simplifyKeyPathConstraint(
     }
   }
 
+  if (keyPathTy->isTypeVariableOrMember()) {
+    return unsolved();
```

Comment:
If I understand it correctly, the reason why this check wasn't here before is related to the fact that it's possible to infer a type of key path based on its components, so key path is always a type variable (unless there is a contextual type) until loop below determines what to bind it too... All of the examples you have added would have a contextual type for a whole key path either from parameter or a coercion. 

Consider this example:

```swift
struct S {
  var bar: String
}

func foo<T>(_: T) {}

foo(\S.bar)
```

Here `keyPathTy` wouldn't get a type until all of the components resolved and key path capabilities are determined from them.

Reply:
Hmm, we don't actually have any types (`keyPathTy` and `getContextualType(keyPath)`) for the KeyPath in my examples, because when we do `addKeyPathConstraint` in CSGen it calls into `simplifyKeyPathConstraint` with a type variable for the KeyPath type.

Reply:
Oh, I think I know what's happening. In your example, we do add an unsolved constraint but that happens at the very of `simplifyKeyPathConstraint`. But for some reason, we don't end up taking that path in my examples.

Reply:
Yeah, I think that's what needs to change and at least in coercion example type of `as` would be associated with key path type variable so there is a contextual type there, I think in parameter one we would attempt key path to be set to a type before simplifying this constraint...

Reply:
Okay. It looks `tryMatchRootAndValueFromType` doesn't actually fail in the normal sense, it just can't call `matchTypes` using the extracted root and value types from the function or key path type if we have a type variable. So, it returns `true` as a fallback. So, I am now unsure if delaying the call to `tryMatchRootAndValueFromType` will have any effect because it will always return `true` for type variables.

After thinking a bit, here are some possible directions I think we can go (for the `\.self` case):

1. We don't treat identity component as an "unresolved" component (not sure why?). So, for such a case, we just fall through and assume we now have a resolved key path type and we call `matchTypes` which returns success. In reality, we don't have any concrete types, just type vars. So, I thought we should also add an unsolved constraint if `TMF_GenerateConstraints` is set for such a case. We already do that for other components.
2. If `tryMatchRootAndValueFromType` can't extract any types, we add an unsolved constraint before returning `true` as fallback.
3. If root or value types are type variables, add unsolved constraint instead of calling `matchTypes` in the end.

Does any of these sound reasonable to you?

For the coercion case, I'll have to make other adjustments. It seems to have the same issue (root and value types are type variables). It seems like `simplifyKeyPathApplicationConstraint` adds an unsolved constraint as well when these are type variables.

Reply:
I think the root issue here is that key path constraint gets picked up constantly e.g. when something changes in one of the components and gets re-evaluated when it shouldn't be. I mentioned this to @LucianoPAlmeida some time ago while he was working on a similar problem - I think we need to make key path behave the way closures do today - it should be resolved when solver can either find contextual information or can prove that there is none. This way we could also avoid all the problems with `PartialKeyPath` being inferred directly, and simplify key path handling. @LucianoPAlmeida do you remember what was PR where we had this conversion?

Reply:
@xedin Yes it was SR-12390 and the PR was #30764 :)

Reply:
I see. I can certainly give that a try! Were there any issues that you faced when you were trying to add this method?

Reply:
@theblixguy I can't remember having any issues, I just had seen that this would require some refactoring and move things around on the way constraints are generated for key path and on `simplifyKeyPathConstraint` that would be more time consuming and as I mentioned to @xedin back then I had some other SR's on the queue that I wanted to tackle, so I felt it was going to be more productive to work on them :) 
But as far as issues in general I don't remember having any ...

Reply:
Okay, thank you! I’ll give it a shot over the weekend and will get back to you.

Reply:
@xedin you mentioned the idea of using one-way equal constraint kind between components in the old PR and that there could be some kind of special logic around key path capability. Could you elaborate what you had in mind with regards to this?

Reply:
If I remember correctly the idea there was that all of the components have to be evaluated in left-to-right order and key path constraint should only be re-activated once _all_ of the components have been resolved, to enforce that we could use one-way constraint.